File: checks/11-codeowners/check.py
#!/usr/bin/env python3
"""Check: CODEOWNERS file existence and validity."""
from __future__ import annotations

import os
import sys
from pathlib import Path
from typing import Optional, List, Tuple

# Valid CODEOWNERS locations (GitHub standard)
CODEOWNERS_LOCATIONS: List[str] = [
    'CODEOWNERS',
    '.github/CODEOWNERS',
    'docs/CODEOWNERS'
]


def find_codeowners_file(repo_path: Path) -> Optional[Path]:
    """
    Find CODEOWNERS file in standard locations.

    Args:
        repo_path: Path to repository root

    Returns:
        Path to CODEOWNERS file if found, None otherwise
    """
    for location in CODEOWNERS_LOCATIONS:
        file_path = repo_path / location
        if file_path.exists() and file_path.is_file():
            return file_path
    return None


def count_ownership_rules(content: str) -> int:
    """
    Count non-empty, non-comment lines (actual ownership rules).

    Args:
        content: CODEOWNERS file content

    Returns:
        Number of ownership rules found
    """
    rule_count = 0
    for line in content.splitlines():
        stripped = line.strip()
        if stripped and not stripped.startswith('#'):
            rule_count += 1
    return rule_count


def check_codeowners(repo_path: Path) -> Tuple[bool, str]:
    """
    Run CODEOWNERS check on repository.

    Args:
        repo_path: Path to repository root

    Returns:
        Tuple of (passed, message)
    """
    codeowners_file = find_codeowners_file(repo_path)

    if codeowners_file is None:
        return False, "No CODEOWNERS file found\nExpected locations: CODEOWNERS, .github/CODEOWNERS, or docs/CODEOWNERS"

    content = codeowners_file.read_text(errors='ignore')
    rule_count = count_ownership_rules(content)

    if rule_count == 0:
        relative_path = codeowners_file.relative_to(repo_path)
        return False, f"CODEOWNERS file found at {relative_path} but contains no ownership rules\nAdd at least one ownership rule (e.g., '* @team-name')"

    relative_path = codeowners_file.relative_to(repo_path)
    plural = 's' if rule_count != 1 else ''
    return True, f"CODEOWNERS file found: {relative_path} ({rule_count} ownership rule{plural})"


def main() -> None:
    """Main entry point."""
    repo_path = Path(os.environ.get('SCORECARD_REPO_PATH', '.'))
    passed, message = check_codeowners(repo_path)

    if passed:
        print(message)
        sys.exit(0)
    else:
        print(message, file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: checks/12-codeowners-quality/check.js
#!/usr/bin/env node
/* eslint-disable no-console */
// Check: CODEOWNERS Quality (Syntax Validation)

import fs from 'fs';
import path from 'path';

const repoPath = process.env.SCORECARD_REPO_PATH || '.';

// Valid CODEOWNERS locations (GitHub standard)
const codeownersLocations = [
    'CODEOWNERS',
    '.github/CODEOWNERS',
    'docs/CODEOWNERS'
];

// Find CODEOWNERS file
let codeownersPath = null;
let relativePath = null;
for (const location of codeownersLocations) {
    const fullPath = path.join(repoPath, location);
    if (fs.existsSync(fullPath)) {
        codeownersPath = fullPath;
        relativePath = location;
        break;
    }
}

if (!codeownersPath) {
    console.error('No CODEOWNERS file found');
    console.error('This check validates CODEOWNERS syntax - create the file first');
    console.error('(Check 11 validates existence, this check validates syntax quality)');
    process.exit(1);
}

// Read and parse content
const content = fs.readFileSync(codeownersPath, 'utf8');
const lines = content.split('\n');

// Valid owner patterns
// @username, @org/team-name, or email@domain.com
const githubUserPattern = /^@[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?$/;
const githubTeamPattern = /^@[a-zA-Z0-9]([a-zA-Z0-9-]*[a-zA-Z0-9])?\/[a-zA-Z0-9_.-]+$/;
const emailPattern = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

function isValidOwner(owner) {
    return githubUserPattern.test(owner) ||
           githubTeamPattern.test(owner) ||
           emailPattern.test(owner);
}

const errors = [];
const validRules = [];
const allOwners = new Set();

lines.forEach((line, index) => {
    const lineNum = index + 1;
    const trimmed = line.trim();

    // Skip empty lines and comments
    if (!trimmed || trimmed.startsWith('#')) {
        return;
    }

    // Split line into parts (pattern + owners)
    const parts = trimmed.split(/\s+/);

    if (parts.length < 2) {
        errors.push(`Line ${lineNum}: Missing owner(s) for pattern '${parts[0]}'`);
        return;
    }

    const pattern = parts[0];
    const owners = parts.slice(1);

    // Validate each owner
    const invalidOwners = [];
    for (const owner of owners) {
        if (isValidOwner(owner)) {
            allOwners.add(owner);
        } else {
            invalidOwners.push(owner);
        }
    }

    if (invalidOwners.length > 0) {
        errors.push(`Line ${lineNum}: Invalid owner format: ${invalidOwners.join(', ')}`);
        errors.push('  Expected: @username, @org/team-name, or email@domain.com');
    } else {
        validRules.push({ pattern, owners, lineNum });
    }
});

// Report results
if (errors.length > 0) {
    console.error('CODEOWNERS file has syntax errors:');
    console.error('');
    errors.forEach(error => console.error(`  ${error}`));
    console.error('');
    console.error('Example valid CODEOWNERS syntax:');
    console.error('  * @default-owner');
    console.error('  *.js @frontend-team');
    console.error('  /docs/ @docs-team @tech-writers');
    console.error('  src/api/ user@example.com @backend-team');
    process.exit(1);
}

if (validRules.length === 0) {
    console.error('CODEOWNERS file found but contains no valid ownership rules');
    console.error('Add at least one rule in the format: <pattern> <owner1> [<owner2>...]');
    process.exit(1);
}

// Success
console.log(`CODEOWNERS syntax validated: ${relativePath}`);
console.log(`  ${validRules.length} ownership rule${validRules.length !== 1 ? 's' : ''}`);
console.log(`  ${allOwners.size} unique owner${allOwners.size !== 1 ? 's' : ''}`);

// Show first few rules as examples
const preview = validRules.slice(0, 3);
preview.forEach(rule => {
    console.log(`  - ${rule.pattern} -> ${rule.owners.join(', ')}`);
});
if (validRules.length > 3) {
    console.log(`  ... and ${validRules.length - 3} more`);
}

process.exit(0);
